app/models/preserved_copy.rb

```diff
@@ -32,14 +32,31 @@ class PreservedCopy < ApplicationRecord
   validates :status, inclusion: { in: statuses.keys }
   validates :version, presence: true
 
-  scope :least_recent_version_audit, lambda { |last_checked_b4_date, storage_dir|
+  scope :by_endpoint_name, lambda { |endpoint_name|
+    joins(:endpoint).where(endpoints: { endpoint_name: endpoint_name })
+  }
+
+  scope :by_storage_location, lambda { |storage_dir|
+    joins(:endpoint).where(endpoints: { storage_location: storage_dir })
+  }
+
+  scope :least_recent_version_audit, lambda { |last_checked_b4_date|
     last_checked_b4_date = normalize_date(last_checked_b4_date)
-    joins(:endpoint)
-      .where(endpoints: { storage_location: storage_dir })
-      .where('last_version_audit IS NULL or last_version_audit < ?', last_checked_b4_date)
+    where('last_version_audit IS NULL or last_version_audit < ?', last_checked_b4_date)
       .order('last_version_audit IS NOT NULL, last_version_audit ASC')
   }
 
+  scope :fixity_check_expired, lambda {
+    joins(:preserved_object)
+      .joins(
+        'INNER JOIN preservation_policies'\
+        ' ON preservation_policies.id = preserved_objects.preservation_policy_id'\
+        ' AND (last_checksum_validation + (fixity_ttl * interval \'1 second\')) < CURRENT_TIMESTAMP'\
```

Comment:
double checking:  "interval" is a reserved word for scope?   And this is checking that the last_checksum_validation is at least 1 second older than the current timestamp, or it is null?  And we're sure it's pp AND (a OR b) rather than  (pp AND a) OR b ?

Reply:
`interval` is actually just plain SQL (and following my own loose formatting convention, i probably should've capitalized it like the other SQL keywords).  the ActiveRecord `joins` method lets you do a join by either:
* specifying a table to join to (and a hash of field/value criteria on which to join if the join isn't on the obvious foreign key column, of if you'd like the join to be more restrictive than just the FK col).  from which it will magically generate your `join` subclause when building a sql statement.
* specifying raw SQL to use for the join if you want to do something more clever than what ActiveRecord allows you to do by way of a table name and a parameter hash.  so here, i've just written out the entire `INNER JOIN` sub-clause for the join to the `preservation_policies` table.

a bit more info on `interval` (which i'm just using here to allow for some simple date math, making the `fixity_ttl` integer addable to the `last_checksum_validation` timestamp, so that the sum can be compared to the current time):
* https://www.postgresql.org/docs/8.2/static/functions-datetime.html
* https://stackoverflow.com/questions/5879448/interval-function-in-oracle#5880358

Reply:
cc'ing @atz here, since he's got some ActiveRecord experience.  if there is a way to do this that generates sane sql and doesn't resort to writing the sql ourselves, i'd be happy to do that instead, since that might be more maintainable in the long run.  i'm not super experience with ActiveRecord, so while i felt like i had a reasonable idea of what the SQL for this should look like, i couldn't figure out how to make AR write that SQL for me.

Reply:
you got it working, and tests prove it's working, yes?   I like the idea of capitalizing INTERVAL and apologize for my SQL ignorance.

Reply:
yes and yes and will do and no worries at all (because this stuff is not particularly intuitive despite what rails would like you to believe)